doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+other downsides. `materialized` is code that's entirely under our control; we
+can deploy any necessary changes to `materialized` with whatever frequency is
+necessary.
+
+### Options for evaluation
+
+* DynamoDB
+
+  Fatal flaw: transactions have an observed median latency of 15ms, with tail
+  latencies of 100ms+.
+
+* A PostgreSQL RDS/Aurora instance per customer.
+
+  Fatal flaw: provisioning a new RDS instance takes tens of minutes. Far too
+  slow to support user onboarding.
+
+* A PostgreSQL RDS/Aurora instance per cloud provider region.
+
+  Open concern: PostgreSQL has an extremely strict limit on the maximum number
+  of connections, and we'll need one connection per environment. Connection
+  poolers like PgBouncer only limit this to the extent that the environments
+  are not constantly issuing queries.
+
+  Open concern: PostgreSQL is not particularly performant in `SERIALIZABLE`
+  mode. Perhaps this won't impact us because we only expect transactions to
+  conflict in the unusual and transient case of having two live controller
+  processes.
+
+  Open concern: a single Aurora instance has a limit of 5000 connections.
+
+* A CockroachDB cluster per cloud provider region.
+
+  Open concern: CockroachDB does not provide strict serializability. How does
+  the anomaly that CockroachDB permits (termed a "causal reverse") impact our
+  ability to provide strict serializability?
+
+  Open concern: CockroachDB does not provide single-key linearizability in the
+  case of clock skew. Empirically: clock skew happens!
+
+* A FoundationDB cluster per cloud provider region.
+
+  Open concern: there are no vendors that offer enterprise support contracts.
+
+  Open soncern: there is a 10MB limit on write transactions.
+
+## Proposal
+
+I propose that we use a single multi-AZ Aurora PostgreSQL instance per cloud
+provider region, with a separate PostgreSQL "database" per customer environment.
+
+Aurora PostgreSQL satisfies all of the hard requirements:
+
+* **Resilience to single availability zone outages.** Supported with multi-AZ
+  deployments.
+
+* **Strong consistency.** Supported with the `SERIALIZABLE` isolation mode.
+  This is not the default; we'll need to remember
+
+* **Single-millisecond commit latency.** TODO: verify this.
+
+* **Easy scalability.** Mostly check, via vertical scalability. We should be
+  able to buy a large enough instance to support up to 5000 customers on a
+  single instance.
+
+* **Compatibility with single binary and bring-your-own-cloud deployments.**
+  In the single binary, SQLite exposes a comparable API. In BYOC deployments,
+  customers will have access to hosted PostgreSQL on all major cloud platforms.
+
+* **Environment schema isolation.** Satisifed by having separate databases
+  per environment.
+
+* **Security sandboxing.** Satisfied by having separate users per environment,
+  with a user only granted access to the corresponding environment.
```

Comment:
where environment == their own postgres db?

Reply:
Environment is a cloud term. It's basically an instance of "stuff" for a single customer organization in a single region. In some of the other platform docs, you might see a concept called REGION, which is very similar.